File: 0006-zigzag-conversion/0006-zigzag-conversion.go
func convert(s string, numRows int) string {
    if numRows == 1 || numRows >= len(s) {
        return s
    }
    
    res := []byte{}
    n := len(s)
    cycleLen := 2 * numRows - 2
    
    for i := 0; i < numRows; i++ {
        for j := 0; j < n; j += cycleLen {
            if j + i < n {
                res = append(res, s[j + i])
            }
            if i != 0 && i != numRows - 1 && j + cycleLen - i < n {
                res = append(res, s[j + cycleLen - i])
            }
        }
    }
    
    return string(res)
}

File: 0006-zigzag-conversion/0006-zigzag-conversion.rs
impl Solution {
    pub fn convert(s: String, num_rows: i32) -> String {
        if num_rows == 1 || num_rows as usize >= s.len() {
            return s
        }
        
        let mut res = String::new();
        let n = s.len();
        let cycle_len = 2 * num_rows as usize - 2;
        
        for i in 0..num_rows as usize {
            for j in (0..n).step_by(cycle_len) {
                if j + i < n {
                    res.push(s.chars().nth(j + i).unwrap());
                }
                if i != 0 && i != num_rows as usize - 1 && j + cycle_len - i < n {
                    res.push(s.chars().nth(j + cycle_len - i).unwrap());
                }
            }
        }
        
        res
    }
}
